templates/support/OAuthAccessTokenManager.swift.hbs

```diff
@@ -0,0 +1,126 @@
+//  
+//  {{>frag/generatedBy}}
+//
+
+/** A token manager that will group refresh attempts to prevent concurrent token requests */
+import Foundation
+
+actor OAuthAccessTokenManager {
+    
+    var accessToken: OAuthToken? {
+        didSet {
+            restartRefreshTimer()
+        }
+    }
+    
+    private let refreshTokenURL: URL?
+    private let automaticallyRefreshTokensThisManySecondsEarly: TimeInterval?
+    private var refreshTask: Task<(), Error>?
+    private var refreshTimer: Timer?
+    private let prepareRequest: ((_ request: URLRequest) -> URLRequest)?
+    
+    init(refreshTokenURL: URL?, automaticallyRefreshTokensThisManySecondsEarly: TimeInterval? = nil, prepareRequest: ((_: URLRequest) -> URLRequest)? = nil) {
+        self.automaticallyRefreshTokensThisManySecondsEarly = automaticallyRefreshTokensThisManySecondsEarly
+        self.prepareRequest = prepareRequest
+        self.refreshTokenURL = refreshTokenURL
+    }
+        
+    private func restartRefreshTimer() {
+        refreshTimer?.invalidate()
+        refreshTimer = nil
+        
+        guard 
+            let autoTime = automaticallyRefreshTokensThisManySecondsEarly,
+            let token = accessToken, 
+            let expiresIn = token.expiresIn,
+            expiresIn - autoTime > 0
+        else {
+            return
+        }
+        
+        let timeout = Date(timeIntervalSinceNow: expiresIn - autoTime)
+        refreshTimer = Timer.scheduledTimer(withTimeInterval: expiresIn - autoTime, repeats: false, block: { t in
+            Task {
+                do {
+                    try await self.refreshToken()
+                } catch {
+                    print("An error ocurred in the automatic refresh: \(error)")
+                }
+            }
+        })
+    }
+    
+    func setAccessToken(_ token: OAuthToken) async throws {
```

Comment:
This is a bit weird to have a `setAccessToken` and a visible `accessToken` property isn't it? This one is async... do we override it sometimes to do some async stuff? I think then we should have a `private` setter on the `accessToken` ivar.

Reply:
it's weird but from my understanding a consequence of using an actor. Actors prevent using the normal setter